pkg/store/proxy_heap.go

```diff
@@ -625,6 +656,24 @@ func newEagerRespSet(
 			l.span.SetTag("processed.samples", seriesStats.Samples)
 			l.span.SetTag("processed.bytes", bytesProcessed)
 			l.span.Finish()
+			sort.Slice(l.bufferedResponses, func(i, j int) bool {
```

Comment:
I think you might be right here. I tried this approach at first, but kept getting non-duplicated results. But now I suspect the reason might be that receivers don't actually return sorted data: https://github.com/thanos-io/thanos/blob/main/pkg/store/tsdb.go#L159

@bwplotka can you also ptal?

Reply:
Unless data is sorted somewhere else (I haven't checked), it sounds like everything worked just because of this explicit sorting via sortDedupLabels() 😱 it should definitely be `true` there

Reply:
We also need to sort series between tenants, so merging https://github.com/thanos-io/thanos/pull/5552 would help us remove the buffer sort completely.

Reply:
:+1: I suggest not getting blocked on that other PR and just make Select() returned sorted series. Does it help with your problem if you convert that to `true`?

Reply:
Hm, the way I understand it, setting sorted to `true` will not help because receivers return series for multiple tenants, and the tsdb select is for one tenant only. So we would still need to sort the buffer on this line to make sure series for all tenants are also sorted. However, with the manual benchmarks I did, it seems that even these store-local sorts help reduce latency relative to the global sort on all series. 

I am also fine with proceeding with this change until we use the proxy in place of multi-tsdb. Then we can completely remove the global sort for the lazy retrieval strategy.

Reply:
So I think TSDB.Select(false) is a red herring. I think TSDB always gives sorted results, , just with extra remote reads in Prometheus we extended this interface -so I think we could double check this AND/OR move to true, but I don't think we are affected 🤔 

On top of that with multiple TSDB it's "heap/mergesort" sorted https://github.com/thanos-io/thanos/blob/6eedf7648757ab1895eb39373ecca62329832b2e/pkg/store/multitsdb.go#L289 this depends on invariant that TSDB series are sorted, but given my first point, I think it is.


So - I agree with @GiedriusS  - I wonder if we can experiment without sorting the series here. But I have a bigger problem - what if series between respSet are duplicated?

Reply:
Unfortunately I don't think that's the case. I remember seeing unsorted series coming from receivers, so I added a repro case [here](https://github.com/thanos-io/thanos/pull/5692/files#diff-a054f78467bb0aedda4e0f1f0d8b8e0cd14e2ee6d3aec0f59ba8c29bdf7c7ed5R30-R65) to see what the behavior is.

When calling [Select](https://github.com/thanos-io/thanos/pull/5692/files#diff-a054f78467bb0aedda4e0f1f0d8b8e0cd14e2ee6d3aec0f59ba8c29bdf7c7ed5R60) with `sortSeries=false`, this is the output I see
```
{__name__="foo", c="2"}
{__name__="foo", c="3"}
{__name__="foo", c="1"}
{__name__="foo", b="2"}
{__name__="foo", b="3"}
{__name__="foo", b="1"}
{__name__="foo", a="2"}
{__name__="foo", a="3"}
{__name__="foo", a="1"}
```

And this is the output with `sortSeries=true`
```
{__name__="foo", a="1"}
{__name__="foo", a="2"}
{__name__="foo", a="3"}
{__name__="foo", b="1"}
{__name__="foo", b="2"}
{__name__="foo", b="3"}
{__name__="foo", c="1"}
{__name__="foo", c="2"}
{__name__="foo", c="3"}
```

This is also good news, because it might be why you were not able to make dedup work without global sorting :)

Reply:
hahaha ;p